cpp/src/parquet/page_index_test.cc

```diff
@@ -0,0 +1,260 @@
+              read_page_location.compressed_page_size);
+    ASSERT_EQ(expected_page_location.first_row_index, read_page_location.first_row_index);
+  }
+}
+
+template <typename DType, typename T = typename DType::c_type>
+void TestReadTypedColumnIndex(const std::string& file_name, int column_id,
+                              size_t num_pages, BoundaryOrder::type boundary_order,
+                              const std::vector<size_t>& page_indices,
+                              const std::vector<bool>& null_pages,
+                              const std::vector<T>& min_values,
+                              const std::vector<T>& max_values,
+                              bool has_null_counts = false,
+                              const std::vector<int64_t>& null_counts = {}) {
+  std::string dir_string(parquet::test::get_data_dir());
+  std::string path = dir_string + "/" + file_name;
+  auto reader = ParquetFileReader::OpenFile(path, false);
+  auto file_metadata = reader->metadata();
+
+  // Get column index location to a specific column chunk.
+  const int row_group_id = 0;
+  ASSERT_LT(row_group_id, file_metadata->num_row_groups());
+  ASSERT_LT(column_id, file_metadata->num_columns());
+  auto index_location = file_metadata->RowGroup(row_group_id)
+                            ->ColumnChunk(column_id)
+                            ->GetColumnIndexLocation();
+  ASSERT_TRUE(index_location.has_value());
+
+  // Read serialized column index from the file.
+  std::shared_ptr<::arrow::io::RandomAccessFile> source;
+  PARQUET_ASSIGN_OR_THROW(source, ::arrow::io::ReadableFile::Open(path));
+  PARQUET_ASSIGN_OR_THROW(auto buffer,
+                          source->ReadAt(index_location->offset, index_location->length));
+  PARQUET_THROW_NOT_OK(source->Close());
+
+  // Deserialize column index.
+  auto properties = default_reader_properties();
+  auto descr = file_metadata->schema()->Column(column_id);
+  std::unique_ptr<ColumnIndex> column_index = ColumnIndex::Make(
+      *descr, buffer->data(), static_cast<uint32_t>(buffer->size()), properties);
+  auto typed_column_index = dynamic_cast<TypedColumnIndex<DType>*>(column_index.get());
+  ASSERT_TRUE(typed_column_index != nullptr);
+
+  // Verify only partial data as there are too many pages.
+  ASSERT_EQ(num_pages, column_index->null_pages().size());
+  ASSERT_EQ(has_null_counts, column_index->has_null_counts());
+  ASSERT_EQ(boundary_order, column_index->boundary_order());
+  for (size_t i = 0; i < page_indices.size(); ++i) {
+    size_t page_id = page_indices.at(i);
+    ASSERT_EQ(null_pages.at(i), column_index->null_pages().at(page_id));
+    if (has_null_counts) {
+      ASSERT_EQ(null_counts.at(i), column_index->null_counts().at(page_id));
+    }
+    // min/max values are only meaningful for non-null pages.
+    if (!null_pages.at(i)) {
+      if constexpr (std::is_same_v<T, double>) {
+        ASSERT_DOUBLE_EQ(min_values.at(i), typed_column_index->min_values().at(page_id));
```

Comment:
This doesn't look right. In the latest version of the PR, `min_values` and `max_values` are dense, i.e. they only contain entries for non-null pages. So you shouldn't be looking at `min_values().at(page_id)` but `min_values().at(non_null_page_index)`.

This also means that none of your tests has both null and non-null pages.